### PR TITLE
perf(ivy): replace select instruction with advance

### PR DESCRIPTION
**Pull request by crisbeto**
Replaces the `select` instruction with a new one called `advance`. Instead of the jumping to a specific index, the new instruction goes forward X amount of elements. The advantage of doing this is that it should generate code that compresses better.

### CODE REVIEW

File: packages/core/src/render3/instructions/advance.ts [hunk 1]
```diff
@@ -8,36 +8,46 @@
...
- * ```
- * @param index the index of the item to act on with the following instructions
- *
+  *   if (rf & 1) {
+  *     text(0, 'Hello');
+  *     text(1, 'Goodbye')
+  *     element(2, 'div');
+  *   }
+  *   if (rf & 2) {
+  *     select(2); // Advance twice to the <div>.
```
Comment (devversion) [inline packages/core/src/render3/instructions/advance.ts]: Should that be `advance`?

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 2]
```diff
@@ -1398,12 +1398,12 @@ describe('i18n support in the template compiler', () => {
...
           }
           if (rf & 2) {
-            $r3$.ɵɵselect(2);
+            $r3$.ɵɵadvance(2);
             $r3$.ɵɵi18nExp(ctx.valueB)(ctx.valueC);
             $r3$.ɵɵi18nApply(3);
             $r3$.ɵɵi18nExp(ctx.valueA);
             $r3$.ɵɵi18nApply(1);
-            $r3$.ɵɵselect(7);
+            $r3$.ɵɵadvance(6);
```
Comment (kara) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Shouldn't this be `advance(5)`?

File: packages/core/src/render3/index.ts [hunk 3]
```diff
@@ -98,6 +98,7 @@ export {
   ɵɵreference,
 
   ɵɵselect,
```
Comment (kara) [inline packages/core/src/render3/index.ts]: Can we add a TODO to remove `select`?

File: packages/core/src/render3/instructions/advance.ts (hunk 1 above)
Comment (kara) [inline packages/core/src/render3/instructions/advance.ts]:
```suggestion
  *     advance(2); // Advance twice to the <div>.
```

**Review by kara — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/index.ts (hunk 3 above)
Comment (crisbeto) [reply packages/core/src/render3/index.ts]: Done. The instruction is also marked as deprecated.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts (hunk 2 above)
Comment (crisbeto) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: I think there's something in i18n that goes backwards, but I'm not sure why. Before I put a guard in place that was similar to the one in `select`, I had a couple of places that generated `advance(-1)`.

Comment (kara) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Ah, looks like we found an existing bug. i18n should not be generating `select` statements at all because `i18n` uses indices directly in `i18nApply` (instead of using `selectedIndex`).  

However,  we currently call `updateInstruction` to generate `i18nExp` and `i18nApply`, which has the select incrementer logic built into it by default. We're passing the select logic indices that have nothing to do with selection. For the first `i18nExp`, we pass in 2 for the current element (because it's in `visitElement`) and for the second one, we pass in 1 for the `i18nStart` index (because we're in `i18nEnd`). This is why you're seeing negative deltas - the current index is being set inappropriately.  We should remove the select instructions altogether for i18n.

**Comment by crisbeto** [comment]
Updated to remove the `advance` instruction for i18n @kara. Also added an extra error during compilation so we don't generate negative `advance` instructions.

**Review by kara — APPROVED** [review]
LGTM

File: packages/core/src/render3/instructions/advance.ts [hunk 4]
```diff
@@ -8,36 +8,46 @@
...
+  *   }
+  *  }
+  * ```
+  * @param delta Number of elements to advance forwards by.
+  *
+  * @codeGenApi
+  */
+export function ɵɵadvance(delta: number): void {
+  ngDevMode && assertGreaterThan(delta, 0, 'Can only advance forward');
+  selectIndexInternal(getLView(), getSelectedIndex() + delta, getCheckNoChangesMode());
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/advance.ts]: Since we are adding a global state read here (super-hot code path) it would be great to have profiling results (property binding and interpolation would be mostly impacted, if there is any difference).

Would you mind running those micro-benchmarks and sharing results?

Comment (crisbeto) [reply packages/core/src/render3/instructions/advance.ts]: I don't mind running them. Should I just execute the ts file or is there something I'm supposed to run them through?

Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/advance.ts]: Running instructions are here: https://github.com/angular/angular/tree/master/packages/core/test/render3/perf

tl;dr is that you build a bundle and execute it as a node script taking timing.

Would be interesting to have numbers before and after this change.
Let me know if you need any help with running those node-based benchmarks.

Comment (crisbeto) [reply packages/core/src/render3/instructions/advance.ts]: I ran it against all the benchmarks that use `select`/`advance` and it does seem to have an effect. Weirdly, a couple of the styling ones were a little quicker and I'm not sure why. Also I suppose that we can avoid reading the global state if we change the `setSelectedIndex` call to something like `advanceSelectedIndex`, but we still need to know the exact index in order to execute the hooks.

#### `interpolation`
```
master:
real	0m1.176s
user	0m1.279s
sys	0m0.048s

this pr:
real	0m1.216s
user	0m1.322s
sys	0m0.047s
```

#### `map_based_style_and_class_bindings`
```
master:
real	1m2.455s
user	1m9.642s
sys	0m1.165s

this pr:
real	0m57.368s
user	1m4.280s
sys	0m1.102s
```

#### `property_binding`
```
master:
real	0m1.195s
user	0m1.205s
sys	0m0.032s

this pr:
real	0m1.203s
user	0m1.219s
sys	0m0.032s
```

#### `style_and_class_bindings`
```
master:
real	0m9.733s
user	0m10.388s
sys	0m0.248s

this pr:
real	0m9.671s
user	0m10.363s
sys	0m0.242s
```

#### `style_binding`
```
master:
real	0m6.542s
user	0m7.261s
sys	0m0.229s

this pr:
real	0m6.617s
user	0m7.339s
sys	0m0.233s
```

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[VE presubmit](https://test.corp.google.com/ui#id=OCL:267992633:BASE:267992664:1568039482096:37502b0a)

[Ivy presubmit](http://test/OCL:267992633:BASE:267995737:1568040645365:949e5959)

**Comment by kara** [comment]
merge-assistance: global approval
